parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetInputFormat.java

```diff
@@ -407,33 +407,36 @@ private static void staticAddInputPathRecursively(List<FileStatus> result,
                 + " found for '" + status.getPath() + "'");
       }
       if (cacheEntry != null) {
-        footers.add(cacheEntry.getFooter());
+        footersMap.put(statusWrapper, cacheEntry.getFooter());
       } else {
         missingStatuses.add(status);
         missingStatusesMap.put(status.getPath(), statusWrapper);
       }
     }
     if (Log.DEBUG) {
-      LOG.debug("found " + footers.size() + " footers in cache and adding up "
+      LOG.debug("found " + footersMap.size() + " footers in cache and adding up "
               + "to " + missingStatuses.size() + " missing footers to the cache");
     }
 
-
-    if (missingStatuses.isEmpty()) {
-      return footers;
+    if (!missingStatuses.isEmpty()) {
+      List<Footer> newFooters = getFooters(config, missingStatuses);
+      for (Footer newFooter : newFooters) {
+        // Use the original file status objects to make sure we store a
+        // conservative (older) modification time (i.e. in case the files and
+        // footers were modified and it's not clear which version of the footers
+        // we have)
+        FileStatusWrapper fileStatus = missingStatusesMap.get(newFooter.getFile());
+        footersCache.put(fileStatus, new FootersCacheValue(fileStatus, newFooter));
+        footersMap.put(fileStatus, newFooter);
+      }
     }
 
-    List<Footer> newFooters = getFooters(config, missingStatuses);
-    for (Footer newFooter : newFooters) {
-      // Use the original file status objects to make sure we store a
-      // conservative (older) modification time (i.e. in case the files and
-      // footers were modified and it's not clear which version of the footers
-      // we have)
-      FileStatusWrapper fileStatus = missingStatusesMap.get(newFooter.getFile());
-      footersCache.put(fileStatus, new FootersCacheValue(fileStatus, newFooter));
+    List<Footer> footers = new ArrayList<Footer>(statuses.size());
+    for (FileStatus status : statuses) {
+      FileStatusWrapper fileStatusWrapper = new FileStatusWrapper(status);
```

Comment:
Instead of creating a new FileStatusWrapper for each file here, what about a slightly different way of doing it:
* For footersMap, use a LinkedHashMap to preserve the key insertion order
* Insert the FileStatusWrapper for missing footers pointing to null
* When assembling the list of footers, loop over the map and if the footer is null, look it up in the cache using the existing FileStatusWrapper

That would save allocations as well as so many footersMap lookups, while also making the part where the cache is loaded completely separate from the rest (it could be refactored into a separate method).